patches/tModLoader/Terraria.ModLoader/DoubleJumpHooks.cs

```diff
@@ -0,0 +1,195 @@
+		{
+			var hook = new HookList(ModLoader.Method(func));
+			hooks.Add(hook);
+			return hook;
+		}
+
+		internal static void Add(ModDoubleJump doubleJump)
+		{
+			doubleJump.index = doubleJumps.Count;
+			doubleJumps.Add(doubleJump);
+		}
+
+		internal static void RebuildHooks()
+		{
+			foreach (var hook in hooks) {
+				hook.arr = ModLoader.BuildGlobalHook(doubleJumps, hook.method).Select(p => p.index).ToArray();
+			}
+		}
+
+		internal static void Unload()
+		{
+			doubleJumps.Clear();
+		}
+
+		internal static void SetupPlayer(Player player)
+		{
+			player.modDoubleJumps = doubleJumps.Select(doubleJump => doubleJump.CreateFor(player)).ToArray();
+		}
+
+		private static HookList HookMidJump = AddHook<Action>(d => d.MidJump);
+
+		/// <summary>
+		/// Allows you to create visuals while the jump is happening
+		/// </summary>
+		/// <param name="player">Jumping player</param>
+		public static void MidJump(Player player)
+		{
+			foreach (int index in HookMidJump.arr) {
+				ModDoubleJump doubleJump = player.modDoubleJumps[index];
+				if (doubleJump.IsMidJump) {
+					player.modDoubleJumps[index].MidJump();
+					return; // Only one jump active at a time
+				}
+			}
+		}
+
+		/// <summary>
+		/// Do the jump if a valid <see cref="nextJump"/> is found in SetNextJump
+		/// </summary>
+		/// <param name="player">Jumping player</param>
+		/// <param name="playSound">if left on true, will play the default double jump sound</param>
+		/// <returns>Jump height multiplier</returns>
+		internal static float Jump(Player player, ref bool playSound)
+		{
+			if (nextJump != -1) {
+				ModDoubleJump doubleJump = player.modDoubleJumps[nextJump];
+				nextJump = -1;
+				doubleJump.effect = true;
+				return doubleJump.Jump(ref playSound);
+			}
+			return 1f;
+		}
+
+		/// <summary>
+		/// Resets all modded double jump enable toggles
+		/// </summary>
+		/// <param name="player">Jumping player</param>
+		internal static void ResetAllEnable(Player player)
+		{
+			foreach (ModDoubleJump doubleJump in player.modDoubleJumps) {
+				doubleJump.enable = false;
+			}
+		}
+
+		/// <summary>
+		/// Stops the ability to do modded double jumps
+		/// </summary>
+		/// <param name="player">Jumping player</param>
+		internal static void ResetAllJumps(Player player)
+		{
+			foreach (ModDoubleJump doubleJump in player.modDoubleJumps) {
+				doubleJump.again = false;
+			}
+		}
+
+		/// <summary>
+		/// Resets all modded double jumps' effects while jumping
+		/// </summary>
+		/// <param name="player">Jumping player</param>
+		internal static void ResetAllEffects(Player player)
+		{
+			foreach (ModDoubleJump doubleJump in player.modDoubleJumps) {
+				doubleJump.effect = false;
+			}
+		}
+
+		/// <summary>
+		/// Returns true if atleast one double jump is able to be used
+		/// </summary>
+		/// <param name="player">Jumping player</param>
+		internal static bool AnyJumpAgain(Player player)
+		{
+			foreach (ModDoubleJump doubleJump in player.modDoubleJumps) {
```

Comment:
This should be simplified to `return player.modDoubleJumps.Any(jump => jump.again);`